File: packages/python_backend/app/error_handling/api_error.py
from typing import Any

class ApiError(Exception):
    def __init__(self, status_code: int, message: str, code: str, details: Any = None):
        super().__init__(message)
        self.status_code = status_code
        self.message = message
        self.code = code
        self.details = details

    def to_dict(self) -> dict[str, Any]:
        return {
            "message": self.message,
            "code": self.code,
            "details": self.details
        }

File: packages/shared/src/error/api-error.ts
// this is moved from the @bnk/router package, maybe I should extend it or something?
// but this issue is I want to use this on both the server and the client but I don't want to have to
// install @bnk/router on the client
export class ApiError extends Error {
  public readonly status: number
  public readonly code: string
  public readonly details?: unknown

  constructor(status = 500, message: string, code = 'INTERNAL_ERROR', details?: unknown) {
    super(message)
    this.name = 'ApiError'
    this.status = status
    this.code = code
    this.details = details

    // Maintain proper stack trace (for V8 engines)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, ApiError)
    }
  }
}

// ParseError is defined in domain-error.ts
